<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import sqlalchemy

# для работы с файлом настроек для подключения БД (try для совместимости с версиями младше 3)
try:
    import configparser
except ImportError:
    import ConfigParser as configparser

#для парсинга городов-миллионников со страницы в Википедии
import requests
from bs4 import BeautifulSoup

    
# Если раздражает warning про необходимость использования engine/connection 
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Создаем объект для хранения настроек подключения, и записываем в него настройки из локального файла settings.ini
config_settings = configparser.ConfigParser()
config_settings.read("settings.ini")

# Создаем переменные, для хранения настроек подключения и присваиваем им значения из локального файла-конфигуратора
DBNAME = config_settings["POSTGRESQL"]["DBNAME"]
USER = config_settings["POSTGRESQL"]["USER"]
PASSWORD = config_settings["POSTGRESQL"]["PASSWORD"]
HOST = config_settings["POSTGRESQL"]["HOST"]
PORT = config_settings["POSTGRESQL"]["PORT"]

In [18]:
# через импорт энжина
'''from sqlalchemy import create_engine
connection = create_engine(f'postgresql+psycopg2://{USER}:{PASSWORD}@{HOST}:{PORT}/{DBNAME}')
print(connection)''';

In [7]:
#Через простое указание подключения
connection = f'postgresql+psycopg2://{USER}:{PASSWORD}@{HOST}:{PORT}/{DBNAME}'
print(connection)
#engine = create_engine(f'postgresql+psycopg2://{USER}:{PASSWORD}@{HOST}:{PORT}/{DBNAME}')
#connection = engine.connect()

postgresql+psycopg2://skillfactory:cCkxxLVrDE8EbvjueeMedPKt@84.201.134.129:5432/project_sql


In [13]:
'''# Создаем соединение с нужными параметрами через psycopg2
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)'''

<connection object at 0x000002EAB5D97BF0; dsn: 'user=skillfactory password=xxx dbname=project_sql host=84.201.134.129 port=5432', closed: 0>


# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [9]:
query_3_1 = f'''SELECT 
                    COUNT (vacancies.id)
                FROM 
                    public.vacancies
'''

In [20]:
#count_vacancies = connection.execute("SELECT COUNT (vac.id) FROM public.vacancies as vac")
count_vacancies = (pd.read_sql_query(query_3_1, connection)).values[0][0]
print('Количество вакансий в таблице vacancies равно', count_vacancies)

AttributeError: 'OptionEngine' object has no attribute 'execute'

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [77]:
query_3_2 = f'''SELECT 
                    COUNT (employers.id)
                FROM 
                    public.employers                
'''

In [78]:
count_employers = (pd.read_sql_query(query_3_2, connection)).values[0][0]
print('Количество работодателей в таблице employers равно', count_employers)

Количество работодателей в таблице employers равно 23501


3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [79]:
query_3_3 = f'''SELECT 
                    COUNT (areas.id)
                FROM 
                    public.areas                
'''

In [80]:
count_areas = (pd.read_sql_query(query_3_3, connection)).values[0][0]
print('Количество регионов в таблице areas равно', count_areas)

Количество регионов в таблице areas равно 1362


4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица industries).

In [81]:
query_3_4 = f'''SELECT 
                    COUNT (industries.id)
                FROM 
                    public.industries               
'''

In [82]:
count_industries = (pd.read_sql_query(query_3_4, connection)).values[0][0]
print('Количество сфер деятельности в таблице industries равно', count_industries)

Количество сфер деятельности в таблице industries равно 294


***

#### **Выводы по предварительному анализу данных**

База состоит из 5 таблиц: ***vacancies***, ***areas***, ***employers***, ***industries*** и ***emplouers_industries***. 

В таблице ***vacancies*** 10 столбцов и 49197 строк, содержащих информацию о вакансиях. Эта таблица наиболее динамичная по записям, так как вакансии чаще теряют актуальность или появляются новые, а также она хранит основную часть информации, которая требуется для обучения модели. Но 5 столбцов в ней  содержат текст, с которым моделям сложнее работать. 

Три таблицы (***areas***, ***industries*** и ***employers***) - это таблицы-справочники, хранящие информацию о городах, сферах деятельности и работодателях соответственно. Таблицы ***employers*** и ***areas***  связаны с таблицей ***vacancies*** через специальные идентификаторы. Таблица  ***industries*** связана с таблицей ***employers***  также через специальные идентификаторы, причем так как связь этих таблиц "много-ко-многим" в базе данных есть дополнительная промежуточная таблица ***emplouers_industries***, которая содержит исключительно ключи-идентификаторы, для образования пар "работодатель-сфера деятельности". 

По предварительному анализу данных можно сделать вывод, что для создание модели машинного обучения, которая будет рекомендовать вакансии клиентам агентства, претендующим на позицию Data Scientist база данных ***project_sql*** подходит, так как содержит довольное большое количество вакансий, охватывает 1362 города и 23501 работодателей. Также база данных содержит всю необходимую информацию: сферу деятельности, чтобы определить те, в которых требуется позиция Data Scientist, город размещения вакантной должности, ключевые навыки и границы заработной платы, а также требования к опыту, график и тип трудоустройства. Этого вполне достаточно, чтобы отсеить необходимые вакансии и обучить модель, поэтому можно продолжить более глубокое изучение данных.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [83]:
query_4_1 = f'''SELECT
                    areas.name as area,
                    COUNT(vacancies.id) as cnt
                FROM 
                    public.vacancies as vacancies
                    JOIN public.areas as areas ON areas.id = vacancies.area_id
                GROUP BY areas.id
                ORDER BY cnt DESC
'''

In [84]:
vacancies_by_area = pd.read_sql_query(query_4_1, connection)
display(vacancies_by_area)

area   cnt
0                      Москва  5333
1             Санкт-Петербург  2851
2                       Минск  2112
3                 Новосибирск  2006
4                      Алматы  1892
..                        ...   ...
764                    Кизляр     1
765                    Джизак     1
766                    Эртиль     1
767                  Арсеньев     1
768  Тура (Красноярский край)     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [85]:
query_4_2 = f'''SELECT 
                    COUNT(vacancies.id)
                FROM 
                    public.vacancies as vacancies
                WHERE (vacancies.salary_from IS NOT NULL) OR (vacancies.salary_to IS NOT NULL)                 
'''

In [86]:
not_null_salary = (pd.read_sql_query(query_4_2, connection)).values[0][0]
print('Количество вакансий, где работодатель указал хотя бы одну границу оплаты труда:', not_null_salary)

Количество вакансий, где работодатель указал хотя бы одну границу оплаты труда: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [87]:
query_4_3 = f'''SELECT 
                    AVG(vacancies.salary_from) as salary_from,
                    AVG(vacancies.salary_to) as salary_to
                FROM 
                    public.vacancies as vacancies             
'''

In [88]:
mean_range_salary = pd.read_sql_query(query_4_3, connection)
print('Среднее значение для нижней границы зарплатной вилки:', round(mean_range_salary['salary_from'].values[0],0))
print('Среднее значение для верхней границы зарплатной вилки:', round(mean_range_salary['salary_to'].values[0],0))

Среднее значение для нижней границы зарплатной вилки: 71065.0
Среднее значение для верхней границы зарплатной вилки: 110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [89]:
query_4_4 = f'''SELECT 
                    vacancies.schedule as schedule,
                    vacancies.employment as employment,
                    COUNT(vacancies.id) as cnt
                FROM 
                    public.vacancies as vacancies
                GROUP BY vacancies.schedule, vacancies.employment
                ORDER BY cnt DESC
'''

In [90]:
schedule_emloyment_combination = pd.read_sql_query(query_4_4, connection)
display(schedule_emloyment_combination)

schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [91]:
query_4_5 = f'''SELECT 
                    vacancies.experience as experience,
                    COUNT(vacancies.id) as cnt
                FROM 
                    public.vacancies as vacancies
                WHERE vacancies.experience IS NOT NULL
                GROUP BY vacancies.experience
                ORDER BY cnt ASC
'''

In [92]:
experience_option = pd.read_sql_query(query_4_5, connection)
display(experience_option)

experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

#### **Выводы по детальному анализу вакансий**

Из 1362 регионов действующие вакансии есть только в 769, что составляет примерно 57% от общего количества регионов. При этом 80% вакансий территориально находятся в первых 50 городах (в рейтинге по количеству вакансий), что объяснимо наличием городов-миллионников, в которых больше работодателей. 

Только в 10102 (~20%) вакансиях работодатель указал обе границы оплаты труда, при этом только 24073 (~49%) вакансий имеют информацию хотя бы об одной границе оплаты труда, то есть больше половины вакансий не имеет информации, которая больше всего интересует соискателей, причиной этого может быть политика компании, когда заработную плату назначают по итогам собеседования, если в компанию требуется руководитель, его зарплату не указывают по причине большой разницы в зарплатах рядовых сотрудников и руководства, либо в компании настолько низкая оплата труда, что она не выдержит конкуренции с другими работодателями, либо компания не проводила исследования или не готова обозначить оплату сотрудника по требуемой вакансии. В данном вопросе требуются дополнительные исследования, чтобы понять, для какого количества вакансий, где требуется позиция Data Scientist границы указаны, и подходит ли это количество, чтобы обучить модель. 

Более половины вакансий (35367) предполагают полную занятость и полный рабочий день, на втором месте удаленная работа и полная занятость. Эти виды деятельности чаще встречаются на позициях Data Scientist, можно предположить, что в данных содержится достаточное количество вакансий необходимых для обучения модели, но этот вопрос также требует более глубокого исследования. 

Большая часть вакансий по опыту работы находятся в препределах от 1 года до 6 лет (от 1 года до 3 лет - 26152 и от 3 до 6 лет - 14511), из исследования резюме мы знаем, что медианный опыт работы составляет 100 месяцев (~8 лет), а значит вакансии в нашей базе данных подходят более 80% соискателей, что хорошо для обучения модели.



# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [93]:
query_5_1 = f'''(SELECT 
                    employers.name as employer,
                    COUNT(vacancies.id) as cnt
                FROM 
                    public.vacancies as vacancies
                    JOIN public.employers as employers ON employers.id = vacancies.employer_id
                GROUP BY employers.id
                ORDER BY cnt DESC
                LIMIT 1)
                UNION ALL
                (SELECT 
                    employers.name as employer,
                    COUNT(vacancies.id) as cnt
                FROM 
                    public.vacancies as vacancies
                    JOIN public.employers as employers ON employers.id = vacancies.employer_id
                GROUP BY employers.id
                ORDER BY cnt DESC
                OFFSET 4 LIMIT 1)
'''

In [94]:
employers_1_and_5_places = pd.read_sql_query(query_5_1, connection)
display(employers_1_and_5_places)

employer   cnt
0         Яндекс  1933
1  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [135]:
query_5_2 = f'''SELECT
                    areas.name,
                    COUNT(vacancies.id) as cnt_vacancies,
                    COUNT(DISTINCT employers.id) as cnt_employers
                FROM 
                    public.areas as areas
                    LEFT JOIN public.vacancies as vacancies ON vacancies.area_id = areas.id
                    JOIN public.employers as employers ON employers.area = areas.id
                    WHERE vacancies.id IS NULL
                GROUP BY areas.id
                ORDER BY cnt_vacancies NULLS FIRST, cnt_employers DESC
'''

In [136]:
employers_and_vacancies_by_areas = pd.read_sql_query(query_5_2, connection)
display(employers_and_vacancies_by_areas)

name  cnt_vacancies  cnt_employers
0                Россия              0            410
1             Казахстан              0            207
2    Московская область              0             75
3    Краснодарский край              0             19
4              Беларусь              0             18
..                  ...            ...            ...
181   Витебская область              0              1
182           Ивацевичи              0              1
183            Дубровно              0              1
184       Буда-Кошелево              0              1
185            Сморгонь              0              1

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [97]:
query_5_3 = f'''SELECT
                    employers.name as employers,
                    COUNT(DISTINCT vacancies.area_id) as cnt_areas_of_vacancies
                FROM 
                    public.employers as employers
                    LEFT JOIN public.vacancies as vacancies ON vacancies.employer_id = employers.id
                GROUP BY employers.id
                ORDER BY cnt_areas_of_vacancies DESC
'''

In [98]:
areas_of_employers_vacancies = pd.read_sql_query(query_5_3, connection)
display(areas_of_employers_vacancies)

employers  cnt_areas_of_vacancies
0                      Яндекс                     181
1                  Ростелеком                     152
2                  Спецремонт                     116
3      Поляков Денис Иванович                      88
4                   ООО ЕФИН                       71
...                       ...                     ...
23496              Но Смокинг                       0
23497        МАСТЕРСКАЯ CREED                       0
23498           Dzing Finance                       0
23499                    КОРС                       0
23500             Наука-Связь                       0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [99]:
query_5_4 = f'''SELECT 
                    COUNT(DISTINCT employers.id)
                FROM 
                    public.employers as employers
                    LEFT JOIN public.employers_industries as emp_ind ON emp_ind.employer_id = employers.id
                WHERE emp_ind.industry_id IS NULL
'''

In [100]:
employers_without_industries = (pd.read_sql_query(query_5_4, connection)).values[0][0]
print('Количество работодателей, для которых не указана сфера деятельности:', employers_without_industries)

Количество работодателей, для которых не указана сфера деятельности: 8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [101]:
query_5_5 = f'''SELECT
                    employers.name as name
                FROM 
                    public.employers as employers
                    JOIN public.employers_industries as emp_ind ON emp_ind.employer_id = employers.id
                GROUP BY employers.id
                HAVING COUNT(emp_ind.industry_id)=4
                ORDER BY name
                OFFSET 2 LIMIT 1                
'''

In [102]:
third_employer_with_4_industries = (pd.read_sql_query(query_5_5, connection)).values[0][0]
print('Компания, находящаяся на 3 месте в списке компаний, у которых 4 сферы деятельности:', third_employer_with_4_industries)

Компания, находящаяся на 3 месте в списке компаний, у которых 4 сферы деятельности: 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [103]:
query_5_6 = f'''SELECT
                    COUNT(employers.id)
                FROM 
                    public.employers_industries as emp_ind
                    JOIN public.employers as employers ON employers.id = emp_ind.employer_id
                    JOIN public.industries as industries ON industries.id = emp_ind.industry_id
                WHERE industries.name LIKE 'Разработка программного обеспечения'    
'''

In [104]:
count_employers_by_mask = (pd.read_sql_query(query_5_6, connection)).values[0][0]
print('Количество работодателей, у которых в качестве сферы деятельности указана Разработка программного обеспечения:', 
      count_employers_by_mask
)

Количество работодателей, у которых в качестве сферы деятельности указана Разработка программного обеспечения: 3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [132]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
contents = requests.get(url).text
soup = BeautifulSoup(contents, 'lxml')
million_cities = []
for i in range(2,18):
    city = soup.select(f'#mw-content-text > div > table.standard.sortable > tbody > tr:nth-child({i}) > td:nth-child(2) > a')
    city = (str(city[0])).split('"')
    million_cities.append(city[3])
    i += 1
million_cities = tuple(million_cities)
print(million_cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [106]:
query_5_7 = f'''SELECT
                    areas.name as citi_name,
                    COUNT(vacancies.id) as cnt
                FROM
                    public.vacancies as vacancies
                    JOIN public.employers as employers ON employers.id = vacancies.employer_id
                    AND employers.name = 'Яндекс'
                    JOIN public.areas as areas ON areas.id = vacancies.area_id
                    AND areas.name in {million_cities}
                GROUP BY areas.name
                UNION ALL
                SELECT
                    'Total',
                    COUNT(*) as cnt
                FROM
                    public.vacancies as vacancies
                    JOIN public.employers as employers ON employers.id = vacancies.employer_id
                    AND employers.name = 'Яндекс'
                    JOIN public.areas as areas ON areas.id = vacancies.area_id
                    AND areas.name in {million_cities}
                ORDER BY cnt ASC
'''

In [107]:
yandex_vacancies = pd.read_sql_query(query_5_7, connection)
print('Количество вакансий компании "Яндекс"  в городах-миллионниках: \n', yandex_vacancies)

Количество вакансий компании "Яндекс"  в городах-миллионниках: 
           citi_name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485


***

#### **Выводы по анализу работодателей**

Исследование работодателей показало, что наибольшее количество вакансий  размещено компанией "Яндекс" - 1933 из них четверть вакансий в городах-миллионниках. При этом у компании "Яндекс", самый большой территориальный охват, ее вакансии размещены в 181 регионах, на втором месте по территориальному охвату компания "Ростелеком". 

В первую двойку регионов, в которых нет вакансий, входят регионы: Россия и Казахстан - географически они включают в себя области и округа, то есть представляют из себя множества других регионов этой же таблицы. И поскольку вакансии можно привязать к более точным географическим местоположениям, рекомендуется рассмотреть целесообразность наличия таких регионов, для обучения нашей модели. Вместе с тем, в данных регионах зарегестрировано 617 работодателей, что может объяснятся тем, что крупные компании, которые работают в разных регионах или имеют распределенную сеть филиалов, но головной офис один, в качестве региона регистрации указывают более крупный географический объект. Либо таким образом работодатели пытаются скрыть вакансии от конкурентов или своих сотрудников. 

У 8419 работодателей не указана сфера деятельности, что также объясняется желанием скрыть поиск сотрудников от конкурентов или уже работающих сотрудников, а также причиной этому может быть новая сфера деятельности, которой нет в справочнике ***industries***.

Сфера деятельности "Разработка программного обеспечения" указана у 3553 работодателей в нашей базе данных, что составляет 15% от всех работодателей, интересно будет узнать, долю вакансий этих 15%, а также подходят ли они для рекомендаций на позицию Data Scientist. В любом случае на данном этапе анализа можно сделать вывод, что шестая часть работодателей точно связана с информационными технологиями.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [109]:
# текст запроса

In [110]:
# результат запроса

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [111]:
# текст запроса

In [112]:
# результат запроса

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [113]:
# текст запроса

In [114]:
# результат запроса

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [115]:
# текст запроса

In [116]:
# результат запроса

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [117]:
# текст запроса

In [118]:
# результат запроса

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [119]:
# текст запроса

In [120]:
# результат запроса

In [121]:
#connection.close()

***

In [122]:
# выводы по предметному анализу

# Общий вывод по проекту

In [123]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Из 23501 работодателя только 15082 (~64%) указали свою сферу деятельности, причиной этого может быть то, что сфера деятельности таких компаний указана в названии, например "ЮТейр, Авиакомпания", либо сфера деятельности новая и отсутствует в справочнике вариантов сфер деятельности, либо компания таким образом хочет скрыть от сотрудников или конкурентов данную вакансию. Также это может быть компания с незаконной деятельностью и к вакансиям таких компаний стоит присмотреться внимательнее. В 23820 (~48%) вакансиях регион регистрации работодателя и регион, где требуется сотрудник не совпадают, это скорее всего компании, у которых есть филиалы и сотрудники требуются в филиал компании, а не в головной офис.

### Это нужно будет перенести вниз
5. Количество уникальных регионов, где есть вакансии (таблица vacancies).

In [124]:
query_3_5 = f'''SELECT 
                    COUNT (DISTINCT vacancies.area_id)
                FROM 
                    public.vacancies              
'''

In [125]:
count_vacancies_areas = (pd.read_sql_query(query_3_5, connection)).values[0][0]
print('Количество регионов, в которых есть вакансии равно', count_vacancies_areas)
print(f'В {count_areas-count_vacancies_areas} регионах не имеется вакансий в базе данных.')

Количество регионов, в которых есть вакансии равно 769
В 593 регионах не имеется вакансий в базе данных.


6. Количество вакансий, где регион регистрации работодателя не совпадает с регионом вакансии.

In [126]:
query_3_6 = f'''SELECT 
                    COUNT(vacancies.id)
                FROM 
                    public.vacancies as vacancies
                    JOIN public.employers as employers ON employers.id = vacancies.employer_id
                WHERE vacancies.area_id <> employers.area                  
'''

In [127]:
different_area = (pd.read_sql_query(query_3_6, connection)).values[0][0]
print('Количество вакансий, где регион вакансии не совпадает с регионом регистрации работодателя:', different_area)

Количество вакансий, где регион вакансии не совпадает с регионом регистрации работодателя: 23820


7. Количество вакансий, для которых работодатель не указал верхнюю и нижнюю границы оплаты труда.

In [128]:
query_3_7 = f'''SELECT 
                    COUNT(vacancies.id)
                FROM 
                    public.vacancies as vacancies
                WHERE (vacancies.salary_from IS NULL) AND (vacancies.salary_to IS NULL)                 
'''

In [129]:
null_salary = (pd.read_sql_query(query_3_7, connection)).values[0][0]
print('Количество вакансий, где работодатель не указал ни одной границы оплаты труда:', null_salary)

Количество вакансий, где работодатель не указал ни одной границы оплаты труда: 25124


8. Количество уникальных работодателей, у которых указана сфера деятельности.

In [130]:
query_3_8 = f'''SELECT 
                    COUNT(DISTINCT employer_id)
                FROM 
                    public.employers_industries                
'''

In [131]:
employers_with_industries = (pd.read_sql_query(query_3_8, connection)).values[0][0]
print('Количество работодателей, для которых указана сфера деятельности (хотя бы одна):', employers_with_industries)

Количество работодателей, для которых указана сфера деятельности (хотя бы одна): 15082
